### PR TITLE
Add new filter to allow modifications of cart contents

### LINKED ISSUE #20962: Inconsistent behaviour for the filter `woocommerce_add_cart_item`
# Description
For the filter `woocommerce_add_cart_item`, we're applying it in [`WC_Cart::add_to_cart`](https://github.com/woocommerce/woocommerce/blob/db403b2/includes/class-wc-cart.php#L1082) which is good. However, we're also applying it in [`WC_Cart_Session::populate_cart_from_order`](https://github.com/woocommerce/woocommerce/blob/a4e0ba2/includes/class-wc-cart-session.php#L315).

For Product Add-Ons, our use case is that we're [adding new cart items](https://github.com/woocommerce/woocommerce-product-addons/blob/f70ee41/includes/class-product-addon-cart.php#L426-L432) to the cart contents in order to tie them to an existing item by hooking on the filter. This works fine when we simply add to cart.

Now, the problem happens when we do "Order Again" for a completed order through the frontend.

In `WC_Cart_Session`, the variable `$cart` is not an actual cart object but an array. As a result of this and since we don't have the cart in context, it does not care about what we do in `WC()->cart` during that filter since it will [overwrite it](https://github.com/woocommerce/woocommerce/blob/a4e0ba2/includes/class-wc-cart-session.php#L128) anyway.

# Reproduction steps
Check the [Product-Addons code](https://github.com/woocommerce/woocommerce-product-addons/blob/f70ee41/includes/class-product-addon-cart.php#L426-L432), but if needed I can supply a minimal example.

# Expected behaviour
I see two potential solutions to this:
1. Offer the current cart object in context on the filter `woocommerce_add_cart_item`
2. Instead of populating cart as an array and then writing it to `WC()->cart`, we write directly to `WC()->cart`

### PR DESCRIPTION
**Pull request by bor0**
### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

Closes #20962.

### CODE REVIEW

**Review by roykho — APPROVED** [review]
LGTM

File: includes/class-wc-cart-session.php [hunk 1]
```diff
@@ -130,6 +130,11 @@ public function get_cart_from_session() {
 			}
 		}
 
+		// If it's not empty, it's been already populated by the loop above
```
Comment (claudiulodro) [inline includes/class-wc-cart-session.php]: [Comments need to end in a full stop](https://travis-ci.org/woocommerce/woocommerce/jobs/416000791#L928)

**Review by claudiulodro — CHANGES_REQUESTED** [review]
Looks good. Just a code style issue needs fixing, then this can be merged.

**Comment by bor0** [comment]
@claudiulodro fixed in a1bec3f20. Thanks!

**Review by claudiulodro — APPROVED** [review]

**Pull request merged**